addons/stock/models/stock_rule.py

```diff
@@ -309,21 +309,22 @@ def _get_stock_move_values(self, product_id, product_qty, product_uom, location_
                 if len(partners) == 1:
                     partner = partners
                     move_dest.partner_id = partner
+        to_refund = values.get('to_refund', False)
 
         move_values = {
             'name': name[:2000],
             'company_id': self.company_id.id or self.location_src_id.company_id.id or self.location_id.company_id.id or company_id.id,
             'product_id': product_id.id,
             'product_uom': product_uom.id,
-            'product_uom_qty': qty_left,
+            'product_uom_qty': qty_left if not to_refund else -qty_left,
             'partner_id': partner.id if partner else False,
-            'location_id': self.location_src_id.id,
-            'location_dest_id': location_id.id,
+            'location_id': self.location_src_id.id if not to_refund else location_id.id,
```

Comment:
It creates a bug in multiple steps. Instead use merge=False. No matter atm if it doens't merge the return